src/memo/once.ml

```diff
@@ -0,0 +1,40 @@
+open Stdune
+open Fiber.O
+
+(* CR-someday amokhov: We could introduce another variant to represent the
+   result of fiber evaluation, e.g. [Finished of 'a]. That would allow us to
+   simplify some logic in Memo, specifically, remove [Sample_attempt.t]. *)
+type 'a state =
+  | Not_forced of { must_not_raise : unit -> 'a Fiber.t }
+  | Forced of 'a Fiber.Ivar.t
+
+type 'a t = { mutable state : 'a state }
+
+(* If a given thunk does in fact raise an exception, forcing it will propagate
+   the exception to the first caller, and leave all subsequent callers stuck,
+   forever waiting for the unfilled [Ivar.t]. *)
+let create ~must_not_raise = { state = Not_forced { must_not_raise } }
+
+let force_impl ?on_blocking t =
+  Fiber.of_thunk (fun () ->
+      match t.state with
+      | Forced ivar -> (
+        match on_blocking with
+        | None -> Fiber.Ivar.read ivar >>| Result.ok
+        | Some on_blocking -> (
+          on_blocking () >>= function
```

Comment:
I pushed a fix that uses `Ivar.peek`.

Reply:
Thanks! As discussed offline, we're going to follow up this PR with some more work to clean things up. For now, adding `Ivar.peek` seems right since it prevents unnecessary executions of `on_blocking`, which means less unnecessary cycle detection work.